dlhub_sdk/utils/validation.py

```diff
@@ -0,0 +1,239 @@
+        # warning message is delivered when a boolean is (perhaps) incorrectly considered an integer
+        warnings.warn("Boolean input has been validated as type Integer, this is likely unintended.", ValidationWarning, stacklevel=2)
+
+
+def _validate_list(li: list, db_entry: dict, path: list) -> None:
+    """Recursively validate each of the elements in li against db_entry["item_type"]
+
+    Args:
+        li (list): The list whose items need to have their type validated
+        db_entry (dict): The entry that stores type data for li
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        TypeError: If any of the items' types do not match
+    """
+    path.append(["list", None])
+    for i, item in enumerate(li):
+        path[-1][1] = i
+        validate(item, db_entry["item_type"], path)
+    path.pop()
+
+
+def _validate_tuple(tup: tuple, db_entries: list, path: list) -> None:
+    """Recursively validate each of the elements in tup against the corresponding entry in db_entries
+
+    Args:
+        tup (tuple): The tuple whose items need to have their type validated
+        db_entries (list): The list of dicts that store the metadata for each item in tup
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If the number of entries and tuple elements do not match
+        TypeError: If any of the items' types do not match
+    """
+    if len(tup) != len(db_entries):
+        raise _generate_err(ValueError, path, msg=f"dl.run expected tuple of length {len(db_entries)}, recieved tuple with length {len(tup)}"
+                                                  + "{loc}")
+
+    path.append(["tuple", None])
+    for i, (given, expected) in enumerate(zip(tup, db_entries)):
+        path[-1][1] = i
+        validate(given, expected, path)
+    path.pop()
+
+
+def _validate_ndarray(arr: ndarray, shape: list, db_entry: dict, path: list) -> None:
+    """Compare the shape of arr with shape and recursively validate each of the elements in arr against db_entry["item_type"]
+
+    Args:
+        arr (ndarray): The ndarray whose shape and items need to be validated
+        shape (list): The shape that arr is expected to have
+        db_entry (dict): The entry where type data may or may not be retrieved
+        path (list): List that stores the path through the data to the current point
+    Returns:
+        None
+    Raises:
+        ValueError: If arr.shape does not match shape
+        TypeError: If any of the items' types do not match
+    """
+    shape = tuple(int(x) if x != "None" else x for x in shape)
+    shape_err = _generate_err(ValueError, path, msg=f"dl.run given improper input: expected ndarray.shape = {shape}, received shape: {arr.shape}"
+                                                    + "{loc}")
+
+    if len(arr.shape) != len(shape):
```

Comment:
Leave some line comments. This first block is checking the shape of the array, the second is checking the contents of the array.

Reply:
We want this code to take minimal effort for someone else to understand it. I can attest that this is taking me more effort than I prefer.